File: 0525-contiguous-array/0525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        int n = nums.length;
        HashMap<Integer, Integer> mp = new HashMap<>();
        int sum = 0, maxLen = 0;
        mp.put(0, -1);

        for(int i = 0; i<n; i++) {
            sum += nums[i] == 0 ? -1 : 1;
            if(mp.containsKey(sum)) {
                maxLen = Math.max(maxLen, i - mp.get(sum));
            }
            else mp.put(sum, i);
        }

        return maxLen;
    }
}

File: 0525-contiguous-array/0525-contiguous-array.py
class Solution(object):
    def findMaxLength(self, nums):
        n, sum, max_len = len(nums), 0, 0
        mp = {}
        mp[0] = -1

        for i in range(n):
            sum += 1 if nums[i] == 1 else -1
            if sum in mp:
                max_len = max(max_len, i - mp[sum])
            else:
                mp[sum] = i
        
        return max_len
        